samplers/aws/xray/internal/manifest.go

```diff
@@ -0,0 +1,359 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	crypto "crypto/rand"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/jinzhu/copier"
+
+	"go.opentelemetry.io/contrib/samplers/aws/xray/internal/util"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+)
+
+const defaultInterval = int64(10)
+
+const manifestTTL = 3600
+
+// Manifest represents a full sampling ruleset and provides
+// option for configuring Logger, Clock and xrayClient.
+type Manifest struct {
+	Rules                          []Rule
+	SamplingTargetsPollingInterval time.Duration
+	refreshedAt                    int64
+	xrayClient                     *xrayClient
+	clientID                       *string
+	logger                         logr.Logger
+	Clock                          util.Clock
+	mu                             sync.RWMutex
+}
+
+// NewManifest return manifest object configured with logging, clock and xrayClient.
+func NewManifest(addr string, logger logr.Logger) (*Manifest, error) {
+	// generate client for getSamplingRules and getSamplingTargets API call
+	client, err := newClient(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	// generate clientID for sampling statistics
+	clientID, err := generateClientID()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Manifest{
+		xrayClient:                     client,
+		Clock:                          &util.DefaultClock{},
+		logger:                         logger,
+		SamplingTargetsPollingInterval: 10 * time.Second,
+		clientID:                       clientID,
+	}, nil
+}
+
+// Expired returns true if the manifest has not been successfully refreshed in
+// manifestTTL seconds.
+func (m *Manifest) Expired() bool {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	return m.refreshedAt < m.Clock.Now().Unix()-manifestTTL
+}
+
+// MatchAgainstManifestRules returns a Rule and boolean flag set as true if rule has been match against span attributes, otherwise nil and false
+func (m *Manifest) MatchAgainstManifestRules(parameters sdktrace.SamplingParameters, serviceName string, cloudPlatform string) (*Rule, bool, error) {
+	m.mu.RLock()
```

Comment:
I think we need read lock here reason being when we do this type of assignment `rules := m.Rules` even then the underlying copy would still remain the same meaning `&m.Rules[0]` and `&rules[0]` would still point to the same address. So I think adding read lock here for the life cycle of a function makes the code more safer because manifest might get updated when runtime is executing this logic.

Reply:
In that case something changed in the refactoring since we meant for the entire manifest to be read once for sampling, and be immutable to not need a lock for the entire sampling decision. IIRC we had gotten to that in the previous iteration - my reading of the code was that in this iteration, `m.Rules` is the equivalent of the previous `s.manifest` which we were treating as immutable

Reply:
Yeah, you're right `m.Rules` is immutable but we still need read locks here because we are iterating over rules to see if there is a match and if there is a match then we sample. Meanwhile other thread can update the manifest rules/targets right? So, in order to synchronize that we need read locks here. Also, we can't do `rules := m.Rules` which basically means we are making a shallow copy so we are iterating over the same memory addresses. 

To explain more:

Main thread ---> reads manifest (for sampling and makes a sampling decision) (read op)
     |
     |
     - - - - - - -> updates manifest in one refresh (write op)
     (second thread)

Reply:
`m.Rules is immutable` and `update the manifest rules` seem to be a contradiction. If the rules are immutable, they can't be updated by another thread, the reason we need a lock to read `m.Rules` is memory visibility, but after taking a reference, if the object is immutable than there would be no further need for locking

Reply:
yep we should be okay to follow this suggestion